File: lua/day2/main.lua
local POINTS = { shapes = { rock = 1, paper = 2, scissor = 3 }, outcome = { lose = 0, draw = 3, win = 6 } }

local function getPoints1(left, right)
  local points = POINTS.shapes[right]
  if left == right then
    return points + POINTS.outcome.draw
  end
  if left == "rock" then
    if right == "paper" then
      return points + POINTS.outcome.win
    end -- scissor
    return points + POINTS.outcome.lose
  elseif left == "paper" then
    if right == "rock" then
      return points + POINTS.outcome.lose
    end -- scissor
    return points + POINTS.outcome.win
  end
  -- scissor
  if right == "rock" then
    return points + POINTS.outcome.win
  end -- paper
  return points + POINTS.outcome.lose
end

local function getPoints2(left, outcome)
  if left == "rock" then
    if outcome == "lose" then
      return POINTS.shapes.scissor + POINTS.outcome.lose
    elseif outcome == "draw" then
      return POINTS.shapes.rock + POINTS.outcome.draw
    end -- win
    return POINTS.shapes.paper + POINTS.outcome.win
  elseif left == "paper" then
    if outcome == "lose" then
      return POINTS.shapes.rock + POINTS.outcome.lose
    elseif outcome == "draw" then
      return POINTS.shapes.paper + POINTS.outcome.draw
    end -- win
    return POINTS.shapes.scissor + POINTS.outcome.win
  end
  -- scissor
  if outcome == "lose" then
    return POINTS.shapes.paper + POINTS.outcome.lose
  elseif outcome == "draw" then
    return POINTS.shapes.scissor + POINTS.outcome.draw
  end -- win
  return POINTS.shapes.rock + POINTS.outcome.win
end

local function play(path)
  local ENCODING1 = { A = "rock", B = "paper", C = "scissor", X = "rock", Y = "paper", Z = "scissor" }
  local ENCODING2 = { X = "lose", Y = "draw", Z = "win" }
  local points1, points2 = 0, 0
  for line in io.lines(path) do
    local char_iter = string.gmatch(line, "%a")
    local left, right = char_iter(), char_iter()
    points1 = points1 + getPoints1(ENCODING1[left], ENCODING1[right])
    points2 = points2 + getPoints2(ENCODING1[left], ENCODING2[right])
  end
  return points1, points2
end

local points1, points2 = play("input.txt")
print("Part1:", points1, "Part2:", points2)

File: rust/day2/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};

type Points = HashMap<&'static str, HashMap<&'static str, i32>>;

fn get_points_1(points_record: &Points, left: &str, right: &str) -> i32 {
    let outcome = &points_record["outcome"];
    points_record["shapes"][right]
        + if left == right {
            outcome["draw"]
        } else {
            match (left, right) {
                ("rock", "paper") | ("scissor", "rock") | ("paper", "scissor") => outcome["win"],
                _ => outcome["lose"],
            }
        }
}

fn get_points_2(points_record: &Points, left: &str, outcome: &str) -> i32 {
    let shapes_record = &points_record["shapes"];
    points_record["outcome"][outcome]
        + match (left, outcome) {
            ("rock", "win") | ("scissor", "lose") => shapes_record["paper"],
            ("paper", "win") | ("rock", "lose") => shapes_record["scissor"],
            ("scissor", "win") | ("paper", "lose") => shapes_record["rock"],
            _ => shapes_record[left],
        }
}

fn main() {
    let points_record = HashMap::from([
        (
            "shapes",
            HashMap::from([("rock", 1), ("paper", 2), ("scissor", 3)]),
        ),
        (
            "outcome",
            HashMap::from([("lose", 0), ("draw", 3), ("win", 6)]),
        ),
    ]);
    let encoding_1 = HashMap::from([
        ("A", "rock"),
        ("B", "paper"),
        ("C", "scissor"),
        ("X", "rock"),
        ("Y", "paper"),
        ("Z", "scissor"),
    ]);
    let encoding_2 = HashMap::from([("X", "lose"), ("Y", "draw"), ("Z", "win")]);
    let buffer = BufReader::new(File::open("input.txt").unwrap());
    let (mut points1, mut points2) = (0, 0);
    for line in buffer.lines() {
        let line = line.unwrap();
        let (left, right) = line.split_once(' ').unwrap();
        points1 += get_points_1(&points_record, encoding_1[left], encoding_1[right]);
        points2 += get_points_2(&points_record, encoding_1[left], encoding_2[right]);
    }
    println!("Part1: {} Part2: {}", points1, points2);
}
